File: RDF2HDFS_Parquet/src/sbd/classes/SchemaExtractor.scala
package sbd.classes

import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.io.Serializable
import java.util.ArrayList
import java.util.Arrays
import java.util.HashSet
import java.util.Iterator
import java.util.List
import java.util.Set
import org.apache.spark.SparkConf
import org.apache.spark.api.java.JavaPairRDD
import org.apache.spark.api.java.JavaRDD
import org.apache.spark.api.java.JavaSparkContext
import org.apache.spark.api.java.function.Function
import org.apache.spark.api.java.function.PairFunction
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.Row
import org.apache.spark.sql.RowFactory
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.hive.HiveContext
import org.codehaus.jettison.json.JSONObject
import sbd.model.Triple
import scala.Tuple2
import com.opencsv.CSVReader

class SchemaExtractor extends Serializable {
  @throws(classOf[IOException])
  def fromCSV(path: String, del: String): ArrayList[String] = {
    val reader: CSVReader = new CSVReader(new FileReader(path))
    val header: Array[String] = reader.readNext
    val headerArrayList: ArrayList[String] = new ArrayList[String](Arrays.asList(header))
    return headerArrayList
  }

  @SuppressWarnings(Array("unchecked"))
  @throws(classOf[IOException])
  def fromSemData(path: String, del: String): ArrayList[(String, Iterable[String])] = {
    val sparkConf: SparkConf = new SparkConf().setAppName("JavaSparkSQL").setMaster("local[2]").set("spark.executor.memory", "1g")
    val ctx: JavaSparkContext = new JavaSparkContext(sparkConf)
    val sqlContext: SQLContext = new SQLContext(ctx)
    val lines: JavaRDD[String] = ctx.textFile(path)
    val triples: JavaRDD[Triple] = lines.map(s -> call(s))
    @SuppressWarnings(Array("unchecked", "rawtypes")) val subject_property: JavaPairRDD[String, (String, String)] = triples.mapToPair(trpl -> new Tuple2(trpl.getSubject(), new Tuple2(trpl.getProperty(), trpl.getObject())))
    val groupBySubject: JavaPairRDD[String, Iterable[(String, String)]] = subject_property.groupByKey
    @SuppressWarnings(Array("unchecked", "serial")) val type_s_po: JavaPairRDD[String, (String, Iterable[(String, String)])] = groupBySubject.mapToPair(new PairFunction[(String, Iterable[(String, String)]), String, (String, Iterable[(String, String)])]() {
      @throws(classOf[Exception])
      def call(list: (String, Iterable[(String, String)])): (String, (String, Iterable[(String, String)])) = {
        val p_o: List[(String, String)] = new ArrayList[(String, String)]
        val types: List[String] = new ArrayList[String]
        var property: String = null
        var `object`: String = null
        var tt: (String, String) = null
        var t2: (String, String) = null
        val subject: String = list._1
        val it: Iterator[(String, String)] = list._2.iterator
        while (it.hasNext) {
          tt = it.next
          property = tt._1
          `object` = tt._2
          if (property == null) {
            types.add(`object`)
          }
          else {
            t2 = new (Any, Any)(property, `object`)
            p_o.add(t2)
          }
        }
        val chosen_type: String = types.get(types.size - 1)
        val s_po: (Any, Any) = new (Any, Any)(subject, p_o)
        val result: (String, (String, Iterable[(String, String)])) = new (String, (String, Iterable[(String, String)]))(chosen_type, s_po)
        return result
      }
    })
    val groupByType: JavaPairRDD[String, Iterable[(String, Iterable[(String, String)])]] = type_s_po.groupByKey
    @SuppressWarnings(Array("serial")) val properties_by_type: JavaPairRDD[String, Iterable[String]] = groupByType.mapToPair(new PairFunction[(String, Iterable[(String, Iterable[(String, String)])]), String, Iterable[String]]() {
      def call(a: (String, Iterable[(String, Iterable[(String, String)])])): (String, Iterable[String]) = {
        val uniqueProperties: Set[String] = new HashSet[String]
        val it: Iterator[(String, Iterable[(String, String)])] = a._2.iterator
        while (it.hasNext) {
          val iter: Iterator[(String, String)] = it.next._2.iterator
          while (iter.hasNext) {
            uniqueProperties.add(iter.next._1)
          }
        }
        return new (String, Iterable[String])(a._1, uniqueProperties)
      }
    })
    val list: List[(String, Iterable[(String, Iterable[(String, String)])])] = groupByType.collect
    val i: Iterator[(String, Iterable[(String, Iterable[(String, String)])])] = list.iterator
    while (i.hasNext) {
      var `type`: String = null
      var subject: String = null
      var columns: String = null
      var values: String = null
      var key_valu: String = null
      var n: Int = 0
      `type` = replace(i.next._1)
      val it: Iterator[(String, Iterable[(String, String)])] = i.next._2.iterator
      while (it.hasNext) {
        columns = ""
        n = 0
        var sep: String = ""
        subject = it.next._1
        val props: List[String] = new ArrayList[String]
        val vals: List[String] = new ArrayList[String]
        val iter: Iterator[(String, String)] = it.next._2.iterator
        while (iter.hasNext) {
          if (n > 0) sep = ","
          n += 1
          val temp: (String, String) = iter.next
          columns += sep + this.replace(temp._1) + " String"
          values += sep + this.replace(temp._2, true)
          key_valu = this.replace(temp._1) + "," + replace(temp._2, true)
          props.add(this.replace(temp._2, true))
        }
        val rdd: JavaRDD[String] = ctx.parallelize(vals)
        val rowRDD: JavaRDD[Row] = rdd.map(new Function[String, Row]() {
          @throws(classOf[Exception])
          def call(record: String): Row = {
            return RowFactory.create(record)
          }
        })
        sqlContext.sql("CREATE TABLE IF NOT EXIST temp " + `type` + " (" + columns + ") STORED AS PARQUET")
        val typeTable: DataFrame = sqlContext.sql("SELECT * FROM temp" + `type` + " LIMIT 0")
        val schemaNestedRDD: DataFrame = sqlContext.createDataFrame(rowRDD, typeTable.schema)
        schemaNestedRDD.registerTempTable(`type`)
        sqlContext.sql("INSERT OVERWRITE INTO TABLE " + `type` + " SELECT * FROM temp" + `type`)
        val tuples: DataFrame = sqlContext.sql("SELECT * FROM " + `type`)
        val result: Array[Row] = tuples.collect
      }
    }
    ctx.close
    return null
  }

  @SuppressWarnings(Array("unchecked"))
  @throws(classOf[IOException])
  def fromSemSchema(path: String, del: String): ArrayList[(String, Iterable[(String, String)])] = {
    val sparkConf: SparkConf = new SparkConf().setAppName("JavaSparkSQL").setMaster("local[2]").set("spark.executor.memory", "1g")
    val ctx: JavaSparkContext = new JavaSparkContext(sparkConf)
    val lines: JavaRDD[String] = ctx.textFile(path)
    val triples: JavaRDD[Triple] = lines.map(line -> {
      String[] parts = line.split(" ");

      Triple triple = new Triple(parts[ 0], parts[ 1], parts[ 2] );
      return triple;
    })
    @SuppressWarnings(Array("serial")) val domainRange: JavaRDD[Triple] = triples.filter(new Function[Triple, Boolean]() {
      @throws(classOf[Exception])
      def call(t: Triple): Boolean = {
        return ((t.getProperty == "<http://www.w3.org/2000/01/rdf-schema#domain>"))
      }
    })
    @SuppressWarnings(Array("rawtypes")) val subject_property: JavaPairRDD[String, (String, String)] = domainRange.mapToPair(trpl -> new Tuple2(trpl.getSubject(), new Tuple2(trpl
      .getProperty(), trpl.getObject())))
    val groupBySubject: JavaPairRDD[String, Iterable[(String, String)]] = subject_property.groupByKey
    val propetiesList: ArrayList[(String, Iterable[(String, String)])] = groupBySubject.collect.asInstanceOf[ArrayList[(String, Iterable[(String, String)])]]
    return propetiesList
  }

  private def call(line: String): Triple = {
    val parts: Array[String] = line.split(" ")
    var triple: Triple = null
    if (parts(1) == "<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>") triple = new Triple(removeTagSymbol(parts(0)), null, removeTagSymbol(parts(2)))
    else triple = new Triple(removeTagSymbol(parts(0)), removeTagSymbol(parts(1)), removeTagSymbol(parts(2)))
    return triple
  }

  private def removeTagSymbol(string: String): String = {
    val str: String = string.replace("<", "").replace(">", "")
    return str
  }

  def JSONtoParquet(jsonFilesPath: String, parquetFilesPath: String, sqlContext: SQLContext) {
    val files: Array[File] = new File(jsonFilesPath).listFiles
    for (file <- files) {
      val fileName: String = file.getName
      val DFFromJSON: DataFrame = sqlContext.jsonFile(jsonFilesPath + fileName)
      DFFromJSON.saveAsParquetFile(parquetFilesPath + fileName.replace(".json", "") + ".parquet")
    }
  }

  private def replace(str: String): String = {
    val newStr: String = str.replace("http://", "p").replace(".", "s").replace("/", "v")
    return newStr
  }

  private def replace(str: String, value: Boolean): String = {
    val newStr: String = str.replace("http://", "p").replace(".", "s").replace("/", "v").replace("^^", "tac")
    return newStr
  }
}

// JavaRDD<Triple> triples = ctx.textFile(path).map(new MapTriples(del));
// JavaRDD<String> words = ctx.textFile(path).flatMap(new MapTuple3(del));
// JavaRDD<String> words = ctx.textFile(path).flatMap(str ->
// Arrays.asList(str.split(" ")));
// DataFrame schemaTriples = sqlContext.createDataFrame(triples,
// Triple.class);
// schemaTriples.groupBy("subject");
// schemaTriples.registerTempTable("triples");
// SQL can be run over RDDs that have been registered as tables.
// DataFrame subjects =
// sqlContext.sql("SELECT COUNT(property) FROM triples GROUP By subject LIMIT 10");
/*groupByType.map(a -> {
String fileName = a._1();
DataFrame peopleFromJsonFile = sqlContext.jsonFile("/home/mohamed/Documents/SBD_Output/Schemas/" + fileName + ".json");
String type = null, columns = null, values = null;
int i = 0;

// Iterate through all the subjects
Iterator<Tuple2<String, Iterable<Tuple2<String, String>>>> it = a._2.iterator();
while (it.hasNext()) {

	// Get the type, which will be the Parquet file name
	type = it.next()._1();
	DataFrame parquetFile = sqlContext.parquetFile("/home/mohamed/Documents/SBD_Output/Schemas/" + type + ".parquet");
	parquetFile.registerTempTable(type);
	columns = "";
	i = 0;
	String sep = "";

	// Iterate through all the (property, object) pairs
	Iterator<Tuple2<String, String>> iter = it.next()._2.iterator();
	while (iter.hasNext()) {
		if(i > 0) sep = ","; i++;
		columns += sep + iter.next()._1();
		values += sep + iter.next()._2();
	}
	sqlContext.sql("INSERT INTO " + type + " (" + columns + ") VALUES (" + values + ")");

	// Read Parquet file

	DataFrame parquetFileFilled = sqlContext.parquetFile("/home/mohamed/Documents/SBD_Output/Schemas/" + type + ".parquet");
	parquetFileFilled.registerTempTable(type);
	DataFrame tuples = sqlContext.sql("SELECT * FROM " + type);
	tuples.collect();
}
return a;
}).collect();*/

File: RDF2HDFS_Parquet/src/sbd/classes/SchemaExtractor.java
package sbd.classes;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStreamWriter;
import java.io.Serializable;
import java.io.Writer;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.hive.HiveContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.codehaus.jettison.json.JSONObject;

import sbd.model.Triple;
import scala.Tuple2;
import scala.Tuple3;

import com.google.common.base.Charsets;
import com.google.common.io.Files;
import com.opencsv.CSVReader;

public class SchemaExtractor implements Serializable {

	private String createQuery;

	public ArrayList<String> fromCSV(String path, String del) throws IOException {

		CSVReader reader = new CSVReader(new FileReader(path));
		String[] header = reader.readNext();

		ArrayList<String> headerArrayList = new ArrayList<String>(Arrays.asList(header));
		return headerArrayList;
	}

	@SuppressWarnings("unchecked")
	public ArrayList<Tuple2<String, Iterable<String>>> fromSemData(String path, String del, String dsName) throws IOException, ClassNotFoundException {

		SparkConf sparkConf = new SparkConf().setAppName("JavaSparkSQL").setMaster("local[2]").set("spark.executor.memory", "4g");
		JavaSparkContext ctx = new JavaSparkContext(sparkConf);
		SQLContext sqlContext = new SQLContext(ctx.sc());

		// Read text file
		JavaRDD<String> lines = ctx.textFile(path);

		// Map lines to Triple objects
		@SuppressWarnings("resource")
		JavaRDD<Triple> triples = lines.map(line -> {
			String[] parts = line.split(del);

			Triple triple = null;
			if (parts[1].equals("<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>"))
				triple = new Triple(replaceInValue(removeTagSymbol(parts[0])), null, replaceInValue(removeTagSymbol(parts[2])));
			else
				triple = new Triple(replaceInValue(removeTagSymbol(parts[0])), replaceInColumn(removeTagSymbol(parts[1])), replaceInValue(removeTagSymbol(parts[2])));
			return triple;
		});

		// Map Triple objects to pairs (Triple.subject,[Triple.property,
		// Triple.object])
		@SuppressWarnings({ "unchecked", "rawtypes" })
		JavaPairRDD<String, Tuple2<String, String>> subject_property = triples.mapToPair(trpl -> new Tuple2(trpl.getSubject(), new Tuple2(trpl
				.getProperty(), trpl.getObject())));

		// Group pairs by subject => s,(p,o)[]
		JavaPairRDD<String, Iterable<Tuple2<String, String>>> groupBySubject = subject_property.groupByKey();

		// Map to pairs (Type,(s(p,o)))
		@SuppressWarnings({ "unchecked", "serial" })
		JavaPairRDD<String, Tuple2<String, Iterable<Tuple2<String, String>>>> type_s_po = groupBySubject.mapToPair(new PairFunction<Tuple2<String, Iterable<Tuple2<String, String>>>, String, Tuple2<String, Iterable<Tuple2<String, String>>>>() {
			@Override
			public Tuple2<String, Tuple2<String, Iterable<Tuple2<String, String>>>> call(Tuple2<String, Iterable<Tuple2<String, String>>> list)
					throws Exception {
				List<Tuple2<String, String>> p_o = new ArrayList<Tuple2<String, String>>();
				List<String> types = new ArrayList<String>();
				String property = null, object = null;
				Tuple2<String, String> tt = null, t2 = null;

				String subject = list._1();
				Iterator<Tuple2<String, String>> it = list._2().iterator();
				while (it.hasNext()) {
					tt = it.next();
					property = tt._1();
					object = tt._2();

					// CAUTION (to be improved): now we are discarding the other types -> We need however to keep them in Columns like "Is also of type"
					if (property == null) {
						types.add(object);
					} else {
						// Form Tuple2(P,O)
						t2 = new Tuple2(property, object);
						p_o.add(t2);
					}
				}

				String chosen_type = types.get(types.size() - 1); // The last type is generally the most specific, but this is definitely not a rule.
																	// We will rather use a hierarchy of classes from the schema (if provided)
				Tuple2 s_po = new Tuple2(subject, p_o);
				Tuple2<String, Tuple2<String, Iterable<Tuple2<String, String>>>> result = new Tuple2<String, Tuple2<String, Iterable<Tuple2<String, String>>>>(chosen_type, s_po);
				return result;
			}
		});

		// Group by type => (type, It(s, It(p, o)))
		JavaPairRDD<String, Iterable<Tuple2<String, Iterable<Tuple2<String, String>>>>> groupByType = type_s_po.groupByKey();
		
		// groupByType: <String, Iterable<Tuple2<String, Iterable<Tuple2<String, String>>>>>
		List<String> keys = groupByType.keys().distinct().collect();

		// Iterate through all types
		int t = 0;
		for (String key : keys) {
			t++;
			if (t < 6){ // To remove later
				@SuppressWarnings("unused")
				JavaRDD<Tuple2<String, Iterable<Tuple2<String, String>>>> rddByKey = getRddByKey(groupByType, key);
	            //rddByKey.take(10).forEach(x -> println(x));
	            
				// Input typeX (s1, (...,...), (...,)
				// Output List<String> = distinct properties
				
							
	            HashMap<String, String> type_columns = new HashMap<String, String>(); // A map to store all columns of one type. We Need it outside the RDD because we need to update it inside the RDD for later use

	            // 1. Iterate through all subjects (RDD lines) and collect the columns (update incrementally the hashmap)
	            JavaRDD<String> cols = rddByKey.flatMap(i -> {
	    			
	    			// 1.1 Iterate through all the (property, object) pairs
	    			Iterator<Tuple2<String, String>> iter = i._2.iterator();
					while (iter.hasNext()) {
						Tuple2<String, String> temp = iter.next(); // To avoid using .next() two times which advances in the iteration
						type_columns.put(temp._1(), ""); // collect columns of one subject (type) in a hashmap (no repetition)
	    			}
					type_columns.put("id", ""); // At last, add the id column
					
					// 1.2 save values to a file
					/*ArrayList<String> writtenColumns = new ArrayList<String>(type_columns.keySet());					
					writeColumns("/home/mohamed/Documents/SBD_Output/TablesColumns/" + replaceInType(key), writtenColumns);
					System.out.println("TYPE" + replaceInType(key));*/
					
					return type_columns.keySet();
					
	    		});
	    		
	    		// Read columns && construct the hashmap
	    		List<String> readColumns = cols.distinct().collect();
	    		
	    		// ArrayList<String> readColumns = readColumns("/home/mohamed/Documents/SBD_Output/TablesColumns/" + replaceInType(key));
	    		for (String j : readColumns) type_columns.put(j,"");
	    		
				// 2 Generate the schema based on the collected properties
				List<StructField> fields_list = new ArrayList<StructField>();				
				for (String s : readColumns) {
					fields_list.add(DataTypes.createStructField(s, DataTypes.StringType, true));
				}
				StructType schema = DataTypes.createStructType(fields_list);				
	    		
				// 3. Map RDD of (subject, Iter(property, object)) to a RDD of Row
	    		JavaRDD<Row> returnValues = rddByKey.map(i -> {
	    			
	    			Row values_list = null;
	    				    			
	    			// 3.1 Iterate through all the (property, object) pairs to save data in the collected columns
	    			Iterator<Tuple2<String, String>> iter = i._2.iterator();
					while (iter.hasNext()) {
						Tuple2<String, String> temp = iter.next(); // To avoid using .next() two times, which advances in the iteration

						// save the value to the appropriate column
						type_columns.put(temp._1(), dsName + ":" + temp._2()); // CAUTION (to improve): this will replace previous values when there are multiple similar properties
					}
					
	    			// 3.2 Get the subject and add it finally to the hashmap
	    			String subject = i._1();
	    			type_columns.put("id", dsName + ":" + subject);
					
					// 3.3 Create the row from the hashmap values
					List<String> vals = new ArrayList<String>(type_columns.values());
					values_list = RowFactory.create(vals.toArray());
					
					// 3.4 Re-initialize the hashmap values --keeping the keys (columns)
					for (String j : type_columns.keySet()) type_columns.put(j,"");

					return values_list;
	    		});
	    		
	    		// 4. Create an RDD by applying a schema to the RDD
				DataFrame typeDataFrame = sqlContext.createDataFrame(returnValues, schema);			
				typeDataFrame.write().parquet("/home/mohamed/Documents/SBD_Output/Parquets/" + replaceInType(key));
			}
        }

		ctx.close();
		return null;
	}

	private JavaRDD getRddByKey(JavaPairRDD<String, Iterable<Tuple2<String, Iterable<Tuple2<String, String>>>>> pairRdd, String key) {
		return pairRdd.filter(v -> v._1().equals(key)).values().flatMap(tuples -> tuples);
	}

	static void returnToZero(HashMap<String, String> mp) {
		Iterator it = mp.entrySet().iterator();
	    while (it.hasNext()) {
	        Map.Entry pair = (Map.Entry)it.next();
	        mp.put(pair.getKey().toString(),"");
	    }
	}

	@SuppressWarnings("unchecked")
	public ArrayList<Tuple2<String, Iterable<Tuple2<String, String>>>> fromSemSchema(String path, String del) throws IOException {
		SparkConf sparkConf = new SparkConf().setAppName("JavaSparkSQL").setMaster("spark://mohamed-N551JK:7077").set("spark.executor.memory", "4g");
		JavaSparkContext ctx = new JavaSparkContext(sparkConf);

		JavaRDD<String> lines = ctx.textFile(path);

		JavaRDD<Triple> triples = lines.map(line -> {
			String[] parts = line.split(" ");

			Triple triple = new Triple(parts[0], parts[1], parts[2]);
			return triple;
		});

		@SuppressWarnings({ "serial" })
		JavaRDD<Triple> domainRange = triples.filter(new Function<Triple, Boolean>() {

			@Override
			public Boolean call(Triple t) throws Exception {
				// TODO Auto-generated method stub
				return (t.getProperty().equals("<http://www.w3.org/2000/01/rdf-schema#domain>"));
			}
		});

		@SuppressWarnings({ "rawtypes" })
		JavaPairRDD<String, Tuple2<String, String>> subject_property = domainRange.mapToPair(trpl -> new Tuple2(trpl.getSubject(), new Tuple2(trpl
				.getProperty(), trpl.getObject())));

		JavaPairRDD<String, Iterable<Tuple2<String, String>>> groupBySubject = subject_property.groupByKey();

		ArrayList<Tuple2<String, Iterable<Tuple2<String, String>>>> propetiesList = (ArrayList<Tuple2<String, Iterable<Tuple2<String, String>>>>) groupBySubject
				.collect();

		return propetiesList;
	}
	
	// Helping methods
	private String removeTagSymbol(String string) {
		String str = string.replace("<", "").replace(">", "");
		return str;
	}
	
	private String replaceInValue(String str) {
		String newStr = str.replace("http://", "");
		return newStr;
	}
	
	private String replaceInType(String str) {
		String newStr = str.replace("/", "__").replace("-", "@");
		return newStr;
	}
	

	private String replaceInColumn(String str) {
		String newStr = str.replace("http://", "");
		return newStr;
	}
	
	
	private void writeColumns(String fileName, ArrayList<String> list) throws IOException {
		FileOutputStream fout = new FileOutputStream (fileName);
		ObjectOutputStream oos = new ObjectOutputStream(fout);
		oos.writeObject(list);
		fout.close();
	}
	
	
	private ArrayList<String> readColumns(String fileName) throws IOException, ClassNotFoundException {
			FileInputStream fin = new FileInputStream (fileName);
			ObjectInputStream oos = new ObjectInputStream(fin);
			ArrayList<String> columns = (ArrayList<String>) oos.readObject();	
			fin.close();
			return columns;
		}		

	}
